### PR TITLE
Modify TGW to unload root table after metadata table on tserver shutdown

### PR DESCRIPTION
**Pull request by dlmarion**
The following tests pass locally for me with these changes.

SslIT.adminStop
SslWithClientAuthIT.adminStop
ShutdownIT.adminStop

### CODE REVIEW

File: server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java [hunk 1]
```diff
@@ -407,6 +407,60 @@ private TableMgmtStats manageTablets(Iterator<TabletManagement> iter,
             state, goal, actions);
       }
 
+      // If the current tablet is hosted and the TabletServer that it's hosted on is being shut
+      // down, then we need to unhost tablets in order. We can't unassign the root tablet before
+      // moving any metadata table tablets that are on this host. We can't unassign metadata
+      // tablets before moving any user table tablets that are on this host.
+      if (current != null && state == TabletState.HOSTED
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:414]: If a tablet server is hosting metadata and user tablet, it should be ok to unload the metadata tablet first if there is another tablet server that can host it.  Once the metadata tablet is loaded elsewhere the the user tablet can unload.  If all tservers are being shutdown, then it seems like the existing code should handle this.

Comment (dlmarion) [reply server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:414]: Yeah, I wasn't sure about the user tables, but I followed the same pattern. For sure, the metadata tablets have to be moved **before** the root tablet. I'll try your suggestion regarding the user tablets.

**Comment by keith-turner** [comment]
Looking at this, I was thinking that the set of servers to shutdown could percolate through the tablet group watchers.   Not sure if it would work.  Was thinking of something like the following.

 * User TGW  uses the managers set of servers to shutdown
 * Metadata TGW uses the User TGW set of servers to shutdown
 * Root TGW uses the Metadata TGW set of servers to shutdown

When a TGW does a full metadata scan it does the following.

 1. tmpshutdownServers = dependentWatcher.getServersToShutdown()
 2. pass tmpshutdownServers to tablet mgmt iterator
 3. do full scan of metadata and remove any servers seen from tmpshutdownServers
 4. set serversToShutdownForParent=tmpshutdownServers

Thinking that TGW would have a method named `getServersToShutdown()` and that it would return `serversToShutdownForParent`.  Also `serversToShutdownForParent` would initially be the empty set and would only be set after TGW does a full scan.

File: server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java (hunk 1 above)
Comment (keith-turner) [reply server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java:414]: I am trying to understand the use case for the new code in this PR.  As long as there is another tablet server to load root or metadata tablet, then it should not matter what order they are unloaded in.  It only matters when all tablet servers are being shutdown, which I thought there was existing code to handle.

**Comment by keith-turner** [comment]
@dlmarion try running the admin stop ITs with the fixes from #3897 and #3898 and see how it goes.

File: server/manager/src/main/java/org/apache/accumulo/manager/Manager.java [hunk 2]
```diff
@@ -632,7 +632,11 @@ TabletGoalState getGoalState(TabletMetadata tm) {
       }
 
       if (tm.hasCurrent() && serversToShutdown.contains(tm.getLocation().getServerInstance())) {
-        return TabletGoalState.SUSPENDED;
+        if (tgw.canSuspendTablets()) {
+          return TabletGoalState.SUSPENDED;
+        } else {
+          return TabletGoalState.UNASSIGNED;
+        }
```
Comment (ctubbsii) [inline server/manager/src/main/java/org/apache/accumulo/manager/Manager.java:639]: Style tip only (feel free to ignore):

```suggestion
        return tgw.canSuspendTablets() ? TabletGoalState.SUSPENDED : TabletGoalState.UNASSIGNED;
```

**Comment by dlmarion** [comment]
> @dlmarion try running the admin stop ITs with the fixes from #3897 and #3898 and see how it goes.

After looking at those issues, I created #3901, which I think should replace #3898 and this PR. #3897 has already been merged, and the tests pass with #3897 and this PR. I'm going to close this.

**Pull request closed without merging**